samza-core/src/main/java/org/apache/samza/startpoint/StartpointManager.java

```diff
@@ -20,87 +20,115 @@
-    Preconditions.checkNotNull(metadataStoreFactory, "MetadataStoreFactory cannot be null");
-    Preconditions.checkNotNull(config, "Config cannot be null");
-    Preconditions.checkNotNull(metricsRegistry, "MetricsRegistry cannot be null");
+  private final MetadataStore metadataStore;
+  private final NamespaceAwareCoordinatorStreamStore fanOutStore;
+  private final NamespaceAwareCoordinatorStreamStore readWriteStore;
+  private final ObjectMapper objectMapper = StartpointObjectMapper.getObjectMapper();
 
-    this.metadataStore = metadataStoreFactory.getMetadataStore(NAMESPACE, config, metricsRegistry);
-    LOG.info("StartpointManager created with metadata store: {}", metadataStore.getClass().getCanonicalName());
-    this.metadataStore.init();
-  }
+  private boolean stopped = true;
 
   /**
    *  Builds the StartpointManager based upon the provided {@link MetadataStore} that is instantiated.
    *  Setting up a metadata store instance is expensive which requires opening multiple connections
-   *  and reading tons of information. Fully instantiated metadata store is taken as a constructor argument
+   *  and reading tons of information. Fully instantiated metadata store is passed in as a constructor argument
    *  to reuse it across different utility classes.
    *
    * @param metadataStore an instance of {@link MetadataStore} used to read/write the start-points.
    */
   public StartpointManager(MetadataStore metadataStore) {
     Preconditions.checkNotNull(metadataStore, "MetadataStore cannot be null");
-    this.metadataStore = new NamespaceAwareCoordinatorStreamStore(metadataStore, NAMESPACE);
+
+    this.metadataStore = metadataStore;
+    this.readWriteStore = new NamespaceAwareCoordinatorStreamStore(metadataStore, NAMESPACE);
+    this.fanOutStore = new NamespaceAwareCoordinatorStreamStore(metadataStore, NAMESPACE_FAN_OUT);
+    LOG.info("Lifecycle of metadata store: {} is managed externally", metadataStore.getClass().getCanonicalName());
+    LOG.info("Startpoints are written to namespace: {} and fanned out to namespace: {} in the metadata store", NAMESPACE,
+        NAMESPACE_FAN_OUT);
   }
 
+  /**
+   * Perform startup operations. Method is idempotent.
+   */
   public void start() {
-    // Metadata store lifecycle is managed outside of the StartpointManager, so not starting it.
+    if (stopped) {
+      LOG.info("starting");
+      readWriteStore.init();
+      fanOutStore.init();
+      stopped = false;
+    } else {
+      LOG.warn("already started");
+    }
   }
 
   /**
-   * Writes a {@link Startpoint} that defines the start position for a {@link SystemStreamPartition}.
-   * @param ssp The {@link SystemStreamPartition} to map the {@link Startpoint} against.
-   * @param startpoint Reference to a Startpoint object.
+   * Perform teardown operations. Method is idempotent.
    */
+  public void stop() {
+    if (!stopped) {
```

Comment:
Is this check necessary? Is start() called after stop() in any legitimate use-cases?

Reply:
Possibly in an external tool or service. Any reason to disallow this?

Reply:
Generally `StartpointManager.stop` is supposed to be called during the end of lifecycle of compoent that manages the `StartpointManager`. Even if `StartpointManager` is stopped, the component that owns it could recreate a new-one rather than invoking start() again. The implicit assumption here is that the metastore implementations can be started after they're stopped, which might not be true.

Reply:
No real use cases I can think of other than a long running service using the StartpointManager wanting to clean up resources when not in use, but I guess they can get around it by creating a new instance.
If we want to prevent the StartpointManager from being usable after a `stop()` then I can remove the `start()` method all together and do the initialization within the constructor only. Long-term, do you think it's ok to impose this restriction?